assets/js/app.js

```diff
@@ -0,0 +1,443 @@
+"use strict";
+
+(async () => {
+
+    // small useful utility functions
+    const saveToLocalStorage = (key, data) => localStorage.setItem(key, JSON.stringify(data));
+    const getFromLocalStorage = key => JSON.parse(localStorage.getItem(key));
+
+    const getSingleCoin = async coinId => getData(`https://api.coingecko.com/api/v3/coins/${coinId}`);
+    const getData = url => fetch(url).then(response => response.json());
+
+    // loader functions
+    const showLoader = () => document.getElementById('loader').style.display = 'block';
+    const hideLoader = () => document.getElementById('loader').style.display = 'none';
+
+    // Store the sixth coin ID for modal interaction (needed global variables)
+    let pendingCoinId = null;
+    let pendingCoinSwitch = null;
+
+    const checkCoinInLocalStorage = coinId => {
+        const coinData = getFromLocalStorage(coinId);
+        if (coinData) {
+            if (new Date().getTime() < coinData.expiration) {
+                // check if current time is smaller than expiration (means less than 2 minutes)
+                return true
+            }
+            return false
+        }
+        return false
+    }
+
+    const saveCoinInfoToStorage = singleCoinData => {
+        const expirationTime = new Date().getTime() + 120 * 1000; // Calculate expiration timestamp (120 seconds)
+        const item = {
+            value: singleCoinData, // The coin data
+            expiration: expirationTime // Expiration timestamp
+        };
+        saveToLocalStorage(singleCoinData.id, item);
+    }
+
+    const initializePopovers = () => {
+        document.querySelectorAll("#coins-container .btn-popover").forEach(button => {
+            button.addEventListener("click", async function () {
+                // Check if a popover instance already exists
+                let popoverInstance = bootstrap.Popover.getInstance(this);
+
```

Comment:
this gets weirder if you search while the popovers are open. they remain displayed on top of cards they no longer represent...
<img width="1310" alt="Screen Shot 2025-01-22 at 19 50 29" src="https://github.com/user-attachments/assets/a14bbf95-6e59-4a8b-a71e-3162caee53fc" />